clang/lib/CIR/CodeGen/CIRGenExpr.cpp

```
@@ -304,6 +305,102 @@ RValue CIRGenFunction::emitAnyExpr(const Expr *e) {
+
+  assert(!cir::MissingFeatures::opCallNoPrototypeFunc());
+  assert(!cir::MissingFeatures::opCallChainCall());
+  assert(!cir::MissingFeatures::hip());
+  assert(!cir::MissingFeatures::opCallMustTail());
+
+  cir::CIRCallOpInterface callOp;
+  RValue callResult =
+      emitCall(funcInfo, callee, &callOp, getLoc(e->getExprLoc()));
+
+  assert(!cir::MissingFeatures::generateDebugInfo());
+
+  return callResult;
+}
+
+CIRGenCallee CIRGenFunction::emitCallee(const clang::Expr *e) {
+  e = e->IgnoreParens();
+
+  // Look through function-to-pointer decay.
+  if (const auto *implicitCast = dyn_cast<ImplicitCastExpr>(e)) {
+    if (implicitCast->getCastKind() == CK_FunctionToPointerDecay ||
+        implicitCast->getCastKind() == CK_BuiltinFnToFnPtr) {
+      return emitCallee(implicitCast->getSubExpr());
+    }
+    // Resolve direct calls.
+  } else if (const auto *declRef = dyn_cast<DeclRefExpr>(e)) {
+    const auto *funcDecl = dyn_cast<FunctionDecl>(declRef->getDecl());
+    assert(
+        funcDecl &&
+        "DeclRef referring to FunctionDecl is the only thing supported so far");
+    return emitDirectCallee(cgm, funcDecl);
+  }
+
+  llvm_unreachable("Nothing else supported yet!");
+}
+
+RValue CIRGenFunction::emitCallExpr(const clang::CallExpr *e) {
+  assert(!cir::MissingFeatures::objCBlocks());
+
+  if (isa<CXXMemberCallExpr>(e)) {
+    cgm.errorNYI(e->getSourceRange(), "call to member function");
+    return RValue::get(nullptr);
+  }
+
+  if (isa<CUDAKernelCallExpr>(e)) {
+    cgm.errorNYI(e->getSourceRange(), "call to CUDA kernel");
+    return RValue::get(nullptr);
+  }
+
+  if (const auto *operatorCall = dyn_cast<CXXOperatorCallExpr>(e)) {
+    if (isa_and_nonnull<CXXMethodDecl>(operatorCall->getCalleeDecl())) {
+      cgm.errorNYI(e->getSourceRange(), "call to member operator");
+      return RValue::get(nullptr);
+    }
+  }
+
+  CIRGenCallee callee = emitCallee(e->getCallee());
+
```

Can you add something like this?
```
if (e->getBuiltinCallee()) {
  cgm.errorNYI(e->getSourceRange(), "Call builtin function");
}
```
I'm concerned with the number of cases that we're silently ignoring in the PR because they aren't implemented yet. I'd like to capture as many as possible with `errorNYI`.

---

Updated.

BTW, I'm curious about the principles of handling NYI during the upstreaming process. In the incubator we typically put an `assert(false)` or a `MissingFeatures` check on the NYI paths. But during upstreaming it seems like:

1. We don't like the `assert(false)` approach apparently.
2. We are still using `MissingFeatures` to track the NYI parts.
3. We now have a new mechanism that emits a diagnostic upon NYI paths.

I'm actually a bit confused about how to choose between 2) and 3). Are there any guidelines or principles? Besides, both of 2) and 3) does not abort normal execution, is it possible that this may break some assumptions dependent by code that follows NYI check points, and eventually lead to clang crashing on strange program points?

---

I think the general approach has been to use MissingFeatures where code will need to be added when a feature is implemented but either (a) the code works in a meaningful way without the feature (for instance, adding an attribute or optimization hint of some kind), or (b) the check for the condition can't be implemented until the feature is implemented (for instance, `isa<_SomeUnimplementedOp_>`.

The fact that compilation continues after `errorNYI` is one of the reasons it is preferred. We want to be able to continue compiling so we can see what problems may lie beyond the unimplemented code. You're right that this will often result in a crash. We're trying to avoid that by returning dummy values of the correct type where possible, but as long as a good diagnostic has been issued, we can live with a crash for now.